lib/common/common/src/budget.rs

```diff
@@ -0,0 +1,359 @@
+                Err(TryAcquireError::NoPermits) => return None,
+                Err(TryAcquireError::Closed) => unreachable!("Cannot acquire CPU permit because CPU budget semaphore is closed, this should never happen"),
+            }
+        } else {
+            None
+        };
+
+        Some((num_cpus as usize, cpu_permit))
+    }
+
+    fn try_acquire_io(&self, desired_io: usize) -> Option<(usize, Option<OwnedSemaphorePermit>)> {
+        let min_required_io = self.min_io_permits(desired_io) as u32;
+        let num_io = self.io_semaphore.available_permits().min(desired_io) as u32;
+        if num_io < min_required_io {
+            return None;
+        }
+
+        let io_permit = if num_io > 0 {
+            let io_result = Semaphore::try_acquire_many_owned(self.io_semaphore.clone(), num_io);
+            match io_result {
+                Ok(permit) => Some(permit),
+                Err(TryAcquireError::NoPermits) => return None,
+                Err(TryAcquireError::Closed) => unreachable!("Cannot acquire IO permit because IO budget semaphore is closed, this should never happen"),
+            }
+        } else {
+            None
+        };
+
+        Some((num_io as usize, io_permit))
+    }
+
+    /// Try to acquire Resources permit for optimization task from global Resource budget.
+    ///
+    /// The given `desired_cpus` is not exact, but rather a hint on what we'd like to acquire.
+    /// - it will prefer to acquire the maximum number of CPUs
+    /// - it will never be higher than the total CPU budget
+    /// - it will never be lower than `min_permits(desired_cpus)`
+    ///
+    /// Warn: only one Resource Permit per thread is allowed. Otherwise, it might lead to deadlocks.
+    ///
+    pub fn try_acquire(&self, desired_cpus: usize, desired_io: usize) -> Option<ResourcePermit> {
+        let (num_cpus, cpu_permit) = self.try_acquire_cpu(desired_cpus)?;
+        let (num_io, io_permit) = self.try_acquire_io(desired_io)?;
+
+        Some(ResourcePermit::new(
+            num_cpus as u32,
+            cpu_permit,
+            num_io as u32,
+            io_permit,
+        ))
+    }
+
+    /// Acquire Resources permit for optimization task from global Resource budget.
+    ///
+    /// This will wait until the required number of permits are available.
+    /// This function is blocking.
+    pub fn acquire(
+        &self,
+        desired_cpus: usize,
+        desired_io: usize,
+        stopped: &AtomicBool,
+    ) -> Option<ResourcePermit> {
+        let mut delay = Duration::from_micros(100);
+        while !stopped.load(std::sync::atomic::Ordering::Relaxed) {
+            if let Some(permit) = self.try_acquire(desired_cpus, desired_io) {
+                return Some(permit);
+            } else {
+                std::thread::sleep(delay);
+                delay = (delay * 2).min(Duration::from_secs(2));
+            }
+        }
+        None
+    }
+
+    pub fn replace_with(
```

Comment:
This can still 'deadlock' if our requested number of CPUs or IO is higher than the total budget/capacity.

Maybe an assertion would suffice here, as we don't suspect to get into that scenario in practice?